kernel/src/path.rs

```diff
@@ -1,136 +1,178 @@
-    url: &'a Url,
-    pub(crate) filename: Option<&'a str>,
-    pub(crate) version: Option<Version>,
-    // if is compacted, this path spans version [`version`, `compacted_to_version`]
-    _compacted_to_version: Option<Version>,
-    pub(crate) is_commit: bool,
-    pub(crate) is_checkpoint: bool,
+#[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+#[cfg_attr(not(feature = "developer-visibility"), visibility::make(pub(crate)))]
+enum LogPathFileType {
+    Commit,
+    SinglePartCheckpoint,
+    #[allow(unused)]
+    UuidCheckpoint(String),
+    // NOTE: Delta spec doesn't actually say, but checkpoint part numbers are effectively 31-bit
+    // unsigned integers: Negative values are never allowed, but Java integer types are always
+    // signed. Approximate that as u32 here.
+    #[allow(unused)]
+    MultiPartCheckpoint {
+        part_num: u32,
+        num_parts: u32,
+    },
+    #[allow(unused)]
+    CompactedCommit {
+        hi: Version,
+    },
+    Unknown,
 }
 
-fn get_filename(path: &str) -> Option<&str> {
-    if path.is_empty() || path.ends_with('/') {
-        None
-    } else {
-        path.rsplit(DELIMITER).next()
+#[derive(Debug)]
+#[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+#[cfg_attr(not(feature = "developer-visibility"), visibility::make(pub(crate)))]
+struct ParsedLogPath<Location: CanTryIntoLogPath = FileMeta> {
+    pub location: Location,
+    pub filename: String,
+    pub extension: String,
+    pub version: Version,
+    pub file_type: LogPathFileType,
+}
+
+// Internal helper used by TryFrom<FileMeta> below. It parses a fixed-length string into the numeric
+// type expected by the caller. A wrong length produces an error, even if the parse succeeded.
+fn parse_path_part<T: FromStr>(
+    value: &str,
+    expect_len: usize,
+    location: &impl CanTryIntoLogPath,
+) -> DeltaResult<T> {
+    match value.parse() {
+        Ok(result) if value.len() == expect_len => Ok(result),
+        _ => Err(Error::invalid_log_path(location)),
     }
 }
 
-fn get_version_opt(version_str_opt: Option<&str>, expected_digits: usize) -> Option<Version> {
-    version_str_opt.and_then(|version_str| {
-        if version_str.len() == expected_digits {
-            version_str.parse().ok()
-        } else {
-            None
-        }
-    })
+// In prod, we ~always construct ParsedLogPath from FileMeta, but in testing it's convenient to use
+// a Url instead. This trait decouples the two.
+#[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+#[cfg_attr(not(feature = "developer-visibility"), visibility::make(pub(crate)))]
+trait CanTryIntoLogPath {
+    fn as_url(&self) -> &Url;
 }
 
-pub(crate) fn version_from_location(location: &Url) -> Option<Version> {
-    let path = location.path();
-    get_filename(path)
-        .and_then(|f| f.split_once('.'))
-        .and_then(|(name, _)| get_version_opt(Some(name), VERSION_LEN))
+impl CanTryIntoLogPath for FileMeta {
```

Comment:
This is a marker trait for `ParsedLogPath::try_from`. I was hoping it would let me work around foreign trait restrictions, but it didn't. I changed it to `AsUrl` tho, simpler name is better.

Reply:
(I also considered just using `AsRef<Url>` but `AsRef` as a general mechanism for exposing struct fields seems weird and error prone, so I decided against it)